### PR TITLE
Discourage reusing Tweens

### LINKED ISSUE #74245: Lack of documentation on impossibility to reuse SceneTreeTween
### Godot version

3.5.1

### System information

Linux

### Issue description

See #71245

The behavior described here is not well documented, it doesn't warn against reusing  SceneTreeTween or what will happen

### Steps to reproduce

_

### Minimal reproduction project

_

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes #74245
Hopefully this will discourage people from even thinking about this 🙄

### CODE REVIEW

File: doc/classes/Tween.xml [hunk 1]
```diff
@@ -89,8 +89,8 @@
...
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not supposed to be reused. Create a new Tween for for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating. While reusing is partially supported by utilizing [method stop], it is [b]very[/b] discouraged and also pointless, as there are no performance benefits.
```
Comment (seppoday) [inline doc/classes/Tween.xml:92]: Double "for"
`Create a new Tween for for each animation and every...`

[the remark this answers is not in the document]
Comment (reptofrog) [reply doc/classes/Tween.xml:92]: I think "pointless" is bit of a strong word, the UI probably shouldn't have an opinion on what's "pointless" and what's not, it's probably best to let the user decide that.

**Comment by follower** [comment]
A pointer for anyone else looking for more context re: `Tween`/`SceneTreeTween` + `pause()`/`stop()` code/behaviour, see:

 * https://github.com/godotengine/godot/pull/71218
 * https://github.com/godotengine/godot/commit/f64dd732b8ac249d052519940e46129960e0d771

File: doc/classes/Tween.xml (hunk 1 above)
Comment (follower) [reply doc/classes/Tween.xml:92]: Perhaps it's for for for emphasis? :)

[the remark this answers is not in the document]
Comment (KoBeWi) [reply doc/classes/Tween.xml:92]: Users are not the ones to decide in this case 🙂
I'm ok with rewording it, but the point is that there are no real benefits.

[the remark this answers is not in the document]
Comment (Mickeon) [reply doc/classes/Tween.xml]: I believe something much closer to the original description, while still heavily noting that reusing them is silly could be suitable.

**Comment by KoBeWi** [comment]
Ok I reworked the PR. Instead of saying "this is somewhat supported, but don't do it", I removed all documentation notes that imply reusing (pretending it's not supported at all) and only left a comment that it results in undefined behavior.

File: doc/classes/Tween.xml [hunk 2]
```diff
@@ -89,8 +89,8 @@
...
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating.
```
Comment (bend-n) [inline doc/classes/Tween.xml:92]:
```suggestion
		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in undefined behavior. Create a new Tween for each animation and start the animation immediately. Tweens start animating immediately, so only create a Tween when you want to start animating.
```

File: doc/classes/Tween.xml [hunk 3]
```diff
@@ -89,8 +89,8 @@
...
+		[b]Note:[/b] Tweens are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
```
Comment (bend-n) [inline doc/classes/Tween.xml:93]:
```suggestion
		[b]Note:[/b] Tweens are processed after all of the nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
```

File: doc/classes/SceneTree.xml [hunk 4]
```diff
@@ -88,7 +88,7 @@
 		<method name="create_tween">
 			<return type="Tween" />
 			<description>
-				Creates and returns a new [Tween].
+				Creates and returns a new [Tween]. The Tween will start automatically on the next process frame or physics frame (depending on [enum TweenProcessMode]).
```
Comment (bend-n) [inline doc/classes/SceneTree.xml:91]:
```suggestion
				Creates and returns a new [Tween]. This Tween will start automatically on the next process or physics frame, depending on [enum TweenProcessMode].
```

File: doc/classes/Node.xml [hunk 5]
```diff
@@ -188,6 +188,7 @@
 				GetTree().CreateTween().BindNode(this);
 				[/csharp]
 				[/codeblocks]
+				The Tween will start automatically on the next process frame or physics frame (depending on [enum TweenProcessMode]).
```
Comment (bend-n) [inline doc/classes/Node.xml:191]:
```suggestion
				This Tween will start automatically on the next process or physics frame, depending on [enum TweenProcessMode].
```

**Review by bend-n — APPROVED** [review]
just saw some english problems

File: doc/classes/Tween.xml (hunk 2 above)
Comment (KoBeWi) [reply doc/classes/Tween.xml:92]: > Create a new Tween for each animation and start the animation immediately.

This repeats in the next sentence and is misleading, as it's not the user who is starting the animation.

Comment (YuriSizov) [reply doc/classes/Tween.xml:92]: Yeah, I don't understand the point of the suggested change.

Comment (YuriSizov) [inline doc/classes/Tween.xml:92]:
```suggestion
		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Keep in mind that Tweens start immediately, so only create a Tween when you want to start animating.
```

Comment (bend-n) [reply doc/classes/Tween.xml:92]: the point is "*in an* undefined behaviour" is strange grammar

**Review by YuriSizov — APPROVED** [review]
Overall, this version seems good to me.

File: doc/classes/Tween.xml (hunk 2 above)
Comment (YuriSizov) [reply doc/classes/Tween.xml:92]: Okay, but you also removed a huge chunk of the note.

Comment (bend-n) [reply doc/classes/Tween.xml]: I did reword it, I just merged it with the repeated sentence after it.

Comment (YuriSizov) [reply doc/classes/Tween.xml]: Your suggestion removes important, if a bit verbose, information. I see no point in merging the two sentences and losing the intended message in the process for both of them.

**Comment by akien-mga** [comment]
Thanks!

**Comment by naturally-intelligent** [comment]
What are best practices for persistent Tweens?

https://godotengine.org/qa/148332/best-practices-for-creating-a-persistent-tween-in-godot-4

I'm kinda stuck because if we are supposed to use AnimationPlayer for this, not sure how to dynamically add Animations at runtime either:

https://godotengine.org/qa/148336/how-can-i-add-animations-to-animationplayer-by-gdscript

**Comment by KoBeWi** [comment]
The answer is in the docs, although probably not obvious for this particular question:
```GDScript
var tween
func animate():
	if tween:
		tween.kill() # Abort the previous animation.
		tween = create_tween()
```
Create the Tween inside `play_effect()` instead of `_ready()`. Still store it in a member variable, but use `kill()` on the old Tween (note that you don't need to check if it's still running).

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.1.

**Pull request merged**